From 568d8b4e88305f4094cb7b5e8226d270020b9318 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Tue, 19 Aug 2014 15:24:28 -0400
Subject: [tools] perf/trace: Beautify send/recv syscall 'flags' arg

Message-id: <1408462094-14194-100-git-send-email-jolsa@redhat.com>
Patchwork-id: 88099
O-Subject: [PATCH RHEL7.1 BZ1131394 099/325] perf trace: Beautify send/recv syscall 'flags' arg
Bugzilla: 1131394
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1131394
https://bugzilla.redhat.com/show_bug.cgi?id=1131394

upstream
========
commit b2cc99fdaad20b6e787f0864a1ac7b14391f9fe6
Author: Arnaldo Carvalho de Melo <acme@redhat.com>
Date: Thu Sep 12 11:54:48 2013 -0300

description
===========
 [root@sandy ~]# perf trace -a -e recvmmsg,recvmsg,recvfrom,sendto,sendmsg,sendmmsg
  6.901 (0.002 ms): 589 recvmsg(fd: 51, msg: 0x7fff35673420, flags: CMSG_CLOEXEC) = -1 EAGAIN Resource temporarily unavailable
  6.966 (0.008 ms): 589 sendmsg(fd: 50, msg: 0x7fff35673230, flags: NOSIGNAL    ) = 961
  6.984 (0.004 ms): 979 sendmsg(fd: 3, msg: 0x7fff5b484940, flags: NOSIGNAL     ) = 945
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/builtin-trace.c b/tools/perf/builtin-trace.c
index 04b632d..f41254f 100644
--- a/tools/perf/builtin-trace.c
+++ b/tools/perf/builtin-trace.c
@@ -319,6 +319,60 @@ static size_t syscall_arg__scnprintf_socket_type(char *bf, size_t size,
 
 #define SCA_SK_TYPE syscall_arg__scnprintf_socket_type
 
+#ifndef MSG_PROBE
+#define MSG_PROBE      0x10
+#endif
+#ifndef MSG_SENDPAGE_NOTLAST
+#define MSG_SENDPAGE_NOTLAST 0x20000
+#endif
+#ifndef MSG_FASTOPEN
+#define MSG_FASTOPEN      0x20000000
+#endif
+
+static size_t syscall_arg__scnprintf_msg_flags(char *bf, size_t size,
+            struct syscall_arg *arg)
+{
+ int printed = 0, flags = arg->val;
+
+ if (flags == 0)
+  return scnprintf(bf, size, "NONE");
+#define P_MSG_FLAG(n) \
+ if (flags & MSG_##n) { \
+  printed += scnprintf(bf + printed, size - printed, "%s%s", printed ? "|" : "", #n); \
+  flags &= ~MSG_##n; \
+ }
+
+ P_MSG_FLAG(OOB);
+ P_MSG_FLAG(PEEK);
+ P_MSG_FLAG(DONTROUTE);
+ P_MSG_FLAG(TRYHARD);
+ P_MSG_FLAG(CTRUNC);
+ P_MSG_FLAG(PROBE);
+ P_MSG_FLAG(TRUNC);
+ P_MSG_FLAG(DONTWAIT);
+ P_MSG_FLAG(EOR);
+ P_MSG_FLAG(WAITALL);
+ P_MSG_FLAG(FIN);
+ P_MSG_FLAG(SYN);
+ P_MSG_FLAG(CONFIRM);
+ P_MSG_FLAG(RST);
+ P_MSG_FLAG(ERRQUEUE);
+ P_MSG_FLAG(NOSIGNAL);
+ P_MSG_FLAG(MORE);
+ P_MSG_FLAG(WAITFORONE);
+ P_MSG_FLAG(SENDPAGE_NOTLAST);
+ P_MSG_FLAG(FASTOPEN);
+ P_MSG_FLAG(CMSG_CLOEXEC);
+#undef P_MSG_FLAG
+
+ if (flags)
+  printed += scnprintf(bf + printed, size - printed, "%s%#x", printed ? "|" : "", flags);
+
+ return printed;
+}
+
+#define SCA_MSG_FLAGS syscall_arg__scnprintf_msg_flags
+
 static size_t syscall_arg__scnprintf_access_mode(char *bf, size_t size,
        struct syscall_arg *arg)
 {
@@ -512,7 +566,12 @@ static struct syscall_fmt {
    .arg_parm  = { [1] = &strarray__rlimit_resources, /* resource */ }, },
  { .name     = "pwrite",     .errmsg = true, .alias = "pwrite64", },
  { .name     = "read",     .errmsg = true, },
- { .name     = "recvfrom",   .errmsg = true, },
+ { .name     = "recvfrom",   .errmsg = true,
+   .arg_scnprintf = { [3] = SCA_MSG_FLAGS, /* flags */ }, },
+ { .name     = "recvmmsg",   .errmsg = true,
+   .arg_scnprintf = { [3] = SCA_MSG_FLAGS, /* flags */ }, },
+ { .name     = "recvmsg",    .errmsg = true,
+   .arg_scnprintf = { [2] = SCA_MSG_FLAGS, /* flags */ }, },
  { .name     = "rt_sigaction", .errmsg = true,
    .arg_scnprintf = { [0] = SCA_SIGNUM, /* sig */ }, },
  { .name     = "rt_sigprocmask", .errmsg = true,
@@ -523,6 +582,12 @@ static struct syscall_fmt {
  { .name     = "rt_tgsigqueueinfo", .errmsg = true,
    .arg_scnprintf = { [2] = SCA_SIGNUM, /* sig */ }, },
  { .name     = "select",     .errmsg = true, .timeout = true, },
+ { .name     = "sendmmsg",    .errmsg = true,
+   .arg_scnprintf = { [3] = SCA_MSG_FLAGS, /* flags */ }, },
+ { .name     = "sendmsg",    .errmsg = true,
+   .arg_scnprintf = { [2] = SCA_MSG_FLAGS, /* flags */ }, },
+ { .name     = "sendto",     .errmsg = true,
+   .arg_scnprintf = { [3] = SCA_MSG_FLAGS, /* flags */ }, },
  { .name     = "setitimer",  .errmsg = true,
    .arg_scnprintf = { [0] = SCA_STRARRAY, /* which */ },
    .arg_parm  = { [0] = &strarray__itimers, /* which */ }, },
-- 
1.7.1